### PR TITLE
Fix  4316 - Error in the character encoding in “User Definition”

### PR DESCRIPTION
**Pull request by gonzalodeniz**
Corrects vowel characters with accents or special characters in the first and last names of users when editing their data.

Error documented in #4316
https://github.com/nilsteampassnet/TeamPass/issues/4316

### CODE REVIEW

File: sources/users.queries.php [hunk 1]
```diff
@@ -1357,8 +1357,8 @@
 
                 $arrData['error'] = false;
                 $arrData['login'] = $rowUser['login'];
-                $arrData['name'] = empty($rowUser['name']) === false && $rowUser['name'] !== NULL ? htmlspecialchars_decode($rowUser['name'], ENT_QUOTES) : '';
-                $arrData['lastname'] = empty($rowUser['lastname']) === false && $rowUser['lastname'] !== NULL ? htmlspecialchars_decode($rowUser['lastname'], ENT_QUOTES) : '';
+                $arrData['name'] = empty($rowUser['name']) === false && $rowUser['name'] !== NULL ? html_entity_decode($rowUser['name'], ENT_QUOTES, 'UTF-8') : '';
+                $arrData['lastname'] = empty($rowUser['lastname']) === false && $rowUser['lastname'] !== NULL ? html_entity_decode($rowUser['lastname'], ENT_QUOTES, 'UTF-8') : '';
```
Comment (corentin-soriano) [inline sources/users.queries.php:1361]: What will happen with this type of payload in the name/lastname?
```js
name<img src=x onerror=alert(1)>
```

**Comment by gonzalodeniz** [comment]
You are right, I think this other solution would prevent the xss attack:

```
$arrData['name'] = empty($rowUser['name']) === false && $rowUser['name'] !== NULL ? htmlspecialchars(mb_convert_encoding($rowUser['name'], 'UTF-8', 'HTML-ENTITIES'), ENT_QUOTES, 'UTF-8') : '';
                $arrData['lastname'] = empty($rowUser['lastname']) === false && $rowUser['lastname'] !== NULL ? htmlspecialchars(mb_convert_encoding($rowUser['lastname'], 'UTF-8', 'HTML-ENTITIES'), ENT_QUOTES, 'UTF-8') : '';
```

I can't right now, but this afternoon I'll try it and if it works I'll update the pull.

Thanks for checking it and letting me know.

**Review by corentin-soriano — CHANGES_REQUESTED** [review]
This code is vulnerable with this payload:
`{"user_id":10000002,"login":"ldap_pwd_change","name":"Ldap","lastname":"name_1234\">&#x3C;img src=x onerror=alert(1)>","email":"test@cs.fr","admin":false,"manager":false,"hr":false,"read_only":false,"personal_folder":true,"new_folder_role_domain":false,"domain":"","isAdministratedByRole":"0","groups":["1"],"allowed_flds":[],"forbidden_flds":[],"action_on_user":"update","form-create-root-folder":false,"mfa_enabled":true}`

image

**Comment by gonzalodeniz** [comment]
How difficult!

I have seen the **strip_tags** function to remove tags. Maybe this will work. I'll check it out this afternoon.

`
$arrData['name'] = empty($rowUser['name']) === false && $rowUser['name'] !== NULL ? htmlspecialchars(strip_tags(mb_convert_encoding($rowUser['name'], 'UTF-8', 'HTML-ENTITIES')), ENT_QUOTES, 'UTF-8') : '';
$arrData['lastname'] = empty($rowUser['lastname']) === false && $rowUser['lastname'] !== NULL ? htmlspecialchars(strip_tags(mb_convert_encoding($rowUser['lastname'], 'UTF-8', 'HTML-ENTITIES')), ENT_QUOTES, 'UTF-8') : '';

`

**Comment by gonzalodeniz** [comment]
I didn't understand the jquery option in this case, sorry.

```
$arrData['name'] = !empty($rowUser['name']) === false && $rowUser['name'] !== NULL ? htmlspecialchars(strip_tags(html_entity_decode($rowUser['name'], ENT_QUOTES | ENT_HTML5, 'UTF-8')), ENT_QUOTES, 'UTF-8') : '';
$arrData['lastname'] = !empty($rowUser['lastname']) === false && $rowUser['lastname'] !== NULL ? htmlspecialchars(strip_tags(html_entity_decode($rowUser['lastname'], ENT_QUOTES | ENT_HTML5, 'UTF-8')), ENT_QUOTES, 'UTF-8') : '';
```

I have been testing and this option does not produce XSS attack and displays correctly the characters á, é, í, ó and ú.

Could you please try it?

Thanks

**Comment by corentin-soriano** [comment]
> I didn't understand the jquery option in this case, sorry.
> 
> ```
> $arrData['name'] = !empty($rowUser['name']) === false && $rowUser['name'] !== NULL ? htmlspecialchars(strip_tags(html_entity_decode($rowUser['name'], ENT_QUOTES | ENT_HTML5, 'UTF-8')), ENT_QUOTES, 'UTF-8') : '';
> $arrData['lastname'] = !empty($rowUser['lastname']) === false && $rowUser['lastname'] !== NULL ? htmlspecialchars(strip_tags(html_entity_decode($rowUser['lastname'], ENT_QUOTES | ENT_HTML5, 'UTF-8')), ENT_QUOTES, 'UTF-8') : '';
> ```
> 
> I have been testing and this option does not produce XSS attack and displays correctly the characters á, é, í, ó and ú.
> 
> Could you please try it?
> 
> Thanks

The `strip_tags` function should not be used in this context:
https://www.php.net/manual/en/function.strip-tags.php
image

The call to `get_user_info` can come from several places.
For consistency with the rest of the project, the data is sent "html safe" to the browser and and we have to convert on the frontend side.

Explanation of my change with jquery:
```js
let data = $('<div>')             // Create "virtual" empty div
                .html(data.login) // Add data.login as html (this will resolve html encoded characters)
                .text()           // Get the text as it is visible
$('#form-login').val(data);       // Put resolved text on the form value
```

**Comment by gonzalodeniz** [comment]
Fixed in v3.1.2.130

**Pull request closed without merging**